File: frontend/src/features/dynamic-simulation/algorithms/thompson.ts
/**
 * Thompson Sampling (Bernoulli) — TypeScript port
 * Reference: Agrawal & Goyal (2012), COLT 2012
 */

import { BaseAlgorithm, SeededRNG } from './base';
import { TransactionContext, ArmState, AlgorithmMetadata, HyperparameterSchema } from '../types';

export class ThompsonSampling implements BaseAlgorithm {
    private arms: string[] = [];
    private alphaPrior = 1.0;
    private betaPrior = 1.0;
    private alpha: Record<string, number> = {};
    private beta: Record<string, number> = {};
    private lastSamples: Record<string, number> = {};
    private lastChosen = '';
    private rng!: SeededRNG;

    initialize(arms: string[], config: Record<string, number>): void {
        this.arms = arms;
        this.alphaPrior = config.alpha_prior ?? 1.0;
        this.betaPrior = config.beta_prior ?? 1.0;
        this.alpha = {};
        this.beta = {};
        for (const arm of arms) {
            this.alpha[arm] = this.alphaPrior;
            this.beta[arm] = this.betaPrior;
        }
        this.lastSamples = {};
        this.lastChosen = '';
        this.rng = new SeededRNG(config.seed ?? 42);
    }

    select(_context: TransactionContext): string {
        const samples: Record<string, number> = {};
        for (const arm of this.arms) {
            samples[arm] = this.rng.betavariate(this.alpha[arm], this.beta[arm]);
        }
        this.lastSamples = samples;
        this.lastChosen = Object.entries(samples).reduce((best, [arm, sample]) =>
            sample > best[1] ? [arm, sample] : best, ['', -Infinity]
        )[0] as string;
        return this.lastChosen;
    }

    update(arm: string, reward: number, _context: TransactionContext): void {
        if (reward === 1) {
            this.alpha[arm] += 1.0;
        } else {
            this.beta[arm] += 1.0;
        }
    }

    getState(): Record<string, ArmState> {
        const state: Record<string, ArmState> = {};
        for (const arm of this.arms) {
            const a = this.alpha[arm];
            const b = this.beta[arm];
            state[arm] = {
                estimatedSR: (a + b) > 0 ? a / (a + b) : null,
                selectionScore: this.lastSamples[arm] ?? null,
                totalSelections: Math.round(a + b - this.alphaPrior - this.betaPrior),
                alpha: Math.round(a * 100) / 100,
                beta: Math.round(b * 100) / 100,
            };
        }
        return state;
    }

    explainLastDecision(): string {
        if (!this.lastChosen) return 'No decision made yet.';
        const sample = this.lastSamples[this.lastChosen] ?? 0;
        const a = this.alpha[this.lastChosen];
        const b = this.beta[this.lastChosen];
        const mean = a / (a + b);
        return `Chose '${this.lastChosen}' with θ=${sample.toFixed(4)} (Beta(α=${a.toFixed(1)}, β=${b.toFixed(1)}), mean=${mean.toFixed(3)})`;
    }

    getHyperparameterSchema(): HyperparameterSchema {
        return {
            alpha_prior: {
                type: 'number',
                default: 1.0,
                min: 0.01,
                max: 100.0,
                description: 'Prior alpha for Beta distribution. 1.0 = uniform prior.',
            },
            beta_prior: {
                type: 'number',
                default: 1.0,
                min: 0.01,
                max: 100.0,
                description: 'Prior beta for Beta distribution. 1.0 = uniform prior.',
            },
        };
    }

    metadata(): AlgorithmMetadata {
        return {
            name: 'Thompson Sampling',
            shortName: 'TS',
            description: 'Bayesian approach sampling from posterior Beta distributions.',
            paper: 'Agrawal & Goyal (2012)',
            paperUrl: 'https://arxiv.org/abs/1111.1797',
            category: 'bandit',
        };
    }
}

File: algorithms/thompson.py
"""
Thompson Sampling (Bernoulli) algorithm.
Reference: Agrawal & Goyal (2012), COLT 2012, arXiv:1111.1797
"""

import random
from typing import List, Dict, Any
from algorithms.base import BaseAlgorithm, TransactionContext


class ThompsonSampling(BaseAlgorithm):
    """
    Thompson Sampling with Beta-Bernoulli model. Samples from
    posterior Beta distributions and selects arm with highest sample.
    """

    def initialize(self, arms: List[str], config: Dict[str, Any]) -> None:
        self.arms = arms
        self.alpha_prior = config.get("alpha_prior", 1.0)
        self.beta_prior = config.get("beta_prior", 1.0)
        self.alpha: Dict[str, float] = {arm: self.alpha_prior for arm in arms}
        self.beta: Dict[str, float] = {arm: self.beta_prior for arm in arms}
        self._last_samples: Dict[str, float] = {}
        self._last_chosen: str = ""
        self._rng = random.Random(config.get("seed", 42))

    def select(self, context: TransactionContext) -> str:
        samples = {}
        for arm in self.arms:
            samples[arm] = self._rng.betavariate(self.alpha[arm], self.beta[arm])

        self._last_samples = samples
        self._last_chosen = max(samples, key=samples.get)
        return self._last_chosen

    def update(self, arm: str, reward: int, context: TransactionContext) -> None:
        if reward == 1:
            self.alpha[arm] += 1.0
        else:
            self.beta[arm] += 1.0

    def get_state(self) -> Dict[str, Any]:
        state = {}
        for arm in self.arms:
            a, b = self.alpha[arm], self.beta[arm]
            total = a + b - 2 * self.alpha_prior  # subtract priors to get observation count
            state[arm] = {
                "estimated_sr": a / (a + b) if (a + b) > 0 else None,
                "selection_score": self._last_samples.get(arm),
                "alpha": round(a, 2),
                "beta": round(b, 2),
                "total_observations": int(total),
            }
        return state

    def explain_last_decision(self) -> str:
        if not self._last_chosen or not self._last_samples:
            return "No decision made yet."
        chosen = self._last_chosen
        sample = self._last_samples.get(chosen, 0)
        a, b = self.alpha[chosen], self.beta[chosen]
        mean = a / (a + b)
        others = {arm: f"{s:.4f}" for arm, s in self._last_samples.items() if arm != chosen}
        return (
            f"Chose '{chosen}' with sampled θ={sample:.4f} "
            f"(Beta(α={a:.1f}, β={b:.1f}), mean={mean:.3f}). "
            f"Other samples: {others}"
        )

    def get_hyperparameter_schema(self) -> Dict[str, Any]:
        return {
            "alpha_prior": {
                "type": "number",
                "default": 1.0,
                "min": 0.01,
                "max": 100.0,
                "description": "Prior alpha parameter for Beta distribution. 1.0 = uniform prior.",
            },
            "beta_prior": {
                "type": "number",
                "default": 1.0,
                "min": 0.01,
                "max": 100.0,
                "description": "Prior beta parameter for Beta distribution. 1.0 = uniform prior.",
            },
        }

    @classmethod
    def metadata(cls) -> Dict[str, str]:
        return {
            "name": "Thompson Sampling",
            "short_name": "TS",
            "description": "Bayesian approach sampling from posterior Beta distributions. Natural exploration-exploitation balance.",
            "paper": "Agrawal & Goyal (2012). Analysis of Thompson Sampling for the Multi-Armed Bandit Problem. COLT 2012.",
            "paper_url": "https://arxiv.org/abs/1111.1797",
            "category": "bandit",
            "non_stationary": "false",
        }
